// @ngscaffolding/models/esm2015/lib/models/inputBuilderModels/inputDetail.model.js
import { BaseEntity } from '../coreModels/baseEntity.model';
export class BaseInput {
}
export var InputTypes;
(function (InputTypes) {
    InputTypes["null"] = "null";
    InputTypes["textbox"] = "textbox";
    InputTypes["number"] = "number";
    InputTypes["email"] = "email";
    InputTypes["password"] = "password";
    InputTypes["textarea"] = "textarea";
    InputTypes["datetime"] = "datetime";
    InputTypes["date"] = "date";
    InputTypes["time"] = "time";
    InputTypes["select"] = "select";
    InputTypes["switch"] = "switch";
    InputTypes["editor"] = "editor";
    InputTypes["listbox"] = "listbox";
    InputTypes["colourpicker"] = "colourpicker";
    InputTypes["spinner"] = "spinner";
    InputTypes["slider"] = "slider";
    InputTypes["checkbox"] = "checkbox";
    InputTypes["tricheckbox"] = "tricheckbox";
    InputTypes["tree"] = "tree";
    InputTypes["selectbutton"] = "selectbutton";
    InputTypes["togglebutton"] = "togglebutton";
    InputTypes["radio"] = "radio";
    InputTypes["radiobuttons"] = "radiobuttons";
    InputTypes["dropdown"] = "dropdown";
    InputTypes["rating"] = "rating";
    InputTypes["autocomplete"] = "autocomplete";
    InputTypes["multiselect"] = "multiselect";
    InputTypes["chips"] = "chips";
    InputTypes["keyfilter"] = "keyfilter";
    InputTypes["fileattach"] = "fileattach";
    InputTypes["dynamic"] = "dynamic";
    InputTypes["search"] = "search";
    InputTypes["repeatform"] = "repeatform";
})(InputTypes || (InputTypes = {}));
export class InputDetail extends BaseEntity {
}
export class InputDetailFile extends InputDetail {
}
export class InputDetailTextBox extends InputDetail {
}
export class InputDetailDateTime extends InputDetail {
}
export class InputDetailReferenceValues extends InputDetail {
}
export class InputDetailDropdown extends InputDetailReferenceValues {
}
export class InputDetailToggleButton extends InputDetail {
}
export class InputDetailTextArea extends InputDetail {
}
export class InputDetailSlider extends InputDetail {
}
export class InputAutoCompete extends InputDetailReferenceValues {
}
export class FileAttach extends InputDetail {
}
export class InputDynamic extends InputDetail {
}
