// @ngscaffolding/models/esm2015/lib/models/coreModels/appSettings.model.js
import { BaseEntity } from './baseEntity.model';
export class AppSettingsValue extends BaseEntity {
}
export var AppSettings;
(function (AppSettings) {
    AppSettings["title"] = "title";
    AppSettings["mobiletitle"] = "mobileTitle";
    AppSettings["iconUrl"] = "iconUrl";
    // Error Handling
    AppSettings["errorShowUser"] = "errorShowUser";
    AppSettings["errorLogConsole"] = "errorLogConsole";
    AppSettings["errorLogServer"] = "errorLogServer";
    AppSettings["pageNotFoundText"] = "pageNotFoundText";
    // Auth Settings
    AppSettings["authSaveinLocalStorage"] = "authSaveinLocalStorage";
    AppSettings["authShowRememberMe"] = "authShowRememberMe";
    AppSettings["authShowForgotPassword"] = "authShowForgotPassword";
    AppSettings["authShowRegister"] = "authShowRegister";
    AppSettings["authTermsAndConditions"] = "authTermsAndConditions";
    AppSettings["authClientId"] = "authClientId";
    AppSettings["authClientSecret"] = "authClientSecret";
    AppSettings["authScope"] = "authScope";
    AppSettings["authUserIdIsEmail"] = "authUserIdIsEmail";
    AppSettings["authUserUniqueField"] = "authUserUniqueField";
    AppSettings["authTokenEndpoint"] = "authTokenEndpoint";
    AppSettings["authPasswordMinLength"] = "authPasswordMinLength";
    AppSettings["authPasswordUpperCase"] = "authPasswordUpperCase";
    AppSettings["authPasswordLowerCase"] = "authPasswordLowerCase";
    AppSettings["authPasswordNumeric"] = "authPasswordNumeric";
    AppSettings["authPasswordSpecial"] = "authPasswordSpecial";
    AppSettings["authRequiredRole"] = "authRequiredRole";
    // OAuth Settings block
    AppSettings["authOAuthSettings"] = "authOAuthSettings";
    // Screen Furniture
    AppSettings["showFooter"] = "showFooter";
    AppSettings["showTopNotifications"] = "showTopNotifications";
    AppSettings["showTopSearch"] = "showTopSearch";
    AppSettings["showTopProfile"] = "showTopProfile";
    AppSettings["showFullMessages"] = "showFullMessages";
    AppSettings["showToastMessages"] = "showToastMessages";
    AppSettings["showProfileSetting"] = "showProfileSetting";
    AppSettings["showProfilePicture"] = "showProfilePicture";
    AppSettings["showUserSetting"] = "showUserSetting";
    AppSettings["menuType"] = "menuType";
    AppSettings["compactMode"] = "compactMode";
    AppSettings["agGridTheme"] = "agGridTheme";
    // Input Builder
    AppSettings["inputShowCalendarIcon"] = "inputShowCalendarIcon";
    AppSettings["inputDateTimeFormat"] = "inputDateTimeFormat";
    AppSettings["inputDateFormat"] = "inputDateFormat";
    AppSettings["inputDateForceUTC"] = "inputDateForceUTC";
    AppSettings["inputDateShowButtonBar"] = "inputDateShowButtonBar";
    // File Uploads
    AppSettings["maximumFileSize"] = "maximumFileSize";
    AppSettings["allowedFileTypes"] = "allowedFileTypes";
    // moment Formats for dates
    AppSettings["dateTimeFormat"] = "dateTimeFormat";
    AppSettings["dateFormat"] = "dateFormat";
    // DataGrid Options
    AppSettings["dataGridIconToolBar"] = "dataGridIconToolBar";
    AppSettings["dataGridAllowConfigureColumns"] = "dataGridAllowConfigureColumns";
    AppSettings["dataGridAllowExport"] = "dataGridAllowExport";
    AppSettings["dataGridAllowSaveView"] = "dataGridAllowSaveView";
    AppSettings["dataGridAllowResetView"] = "dataGridAllowResetView";
    AppSettings["dataGridAllowShareView"] = "dataGridAllowShareView";
    AppSettings["dataGridUseEnterprise"] = "dataGridUseEnterprise";
    // Dashboard Settings
    AppSettings["dashboardDefaultConfig"] = "dashboardDefaultConfig";
    // User Preferences
    AppSettings["prefHideTheme"] = "prefHideTheme";
    AppSettings["apiHome"] = "apiHome";
    AppSettings["apiAuth"] = "apiAuth";
    AppSettings["apiServers"] = "apiServers";
    AppSettings["cacheDefaultSeconds"] = "cacheDefaultSeconds";
    // Mobile settings
    AppSettings["isMobile"] = "isMobile";
    AppSettings["mobileTitle"] = "mobileTitle";
    AppSettings["mobileCacheNames"] = "mobileCacheNames";
    AppSettings["mobileDefaultTimeout"] = "mobileDefaultTimeout";
    AppSettings["mobileDefaultRetries"] = "mobileDefaultRetries";
    AppSettings["mobileDefaultPolling"] = "mobileDefaultPolling";
})(AppSettings || (AppSettings = {}));
